File: ios/Application/Ports/ArchiveEngine.swift
import Foundation

protocol ArchiveEngine: AnyObject {
  func inspect(input: ArchiveInput) async throws -> ArchiveInspection
  func inspect(data: Data) throws -> ArchiveInspection
  func readEntry(session: ArchiveEngineSession, path: String, limit: UInt64, password: String?) async throws -> Data
  func extractEntry(session: ArchiveEngineSession, path: String, to destination: URL, limit: UInt64, password: String?) async throws -> UInt64
  func createArchive(plan: CreationPlan, output: ArchiveOutput, onProgress: @escaping (ProgressSnapshot) -> Void) async throws -> CreationResult
}

struct ArchiveInspection {
  let format: String
  let entryCount: Int
  let entries: [EntryDescriptor]
  let compressedSize: UInt64?
  let totalUncompressedSize: UInt64?
  let encrypted: Bool
  let comment: String?
}

struct EntryDescriptor {
  let index: Int
  let path: String
  let kind: String
  let compressedSize: UInt64
  let uncompressedSize: UInt64
  let encrypted: Bool
  let compressionMethod: String
  let crc32: UInt32?
  let unixMode: UInt16?
}

struct CreationPlan {
  let entries: [EntryInput]
  let compressionProfile: String?
  let compressionLevel: Int?
  let storeAlreadyCompressed: Bool
  let encryptionMethod: String?
  let encryptionPassword: String?
}

struct EntryInput {
  let kind: String
  let sourcePath: String?
  let archivePath: String
  let data: Data?
  let recursive: Bool?
  let includeHidden: Bool?
  let followSymlinks: Bool?
  let compressionMethod: String?
  let modifiedAt: TimeInterval?
}

struct CreationResult {
  let entryCount: Int
  let inputBytes: UInt64
  let outputBytes: UInt64
}

File: android/src/main/java/com/mcodex/nitroarchive/application/ports/ArchiveEngine.kt
package com.mcodex.nitroarchive.application.ports

import com.mcodex.nitroarchive.domain.ProgressSnapshot
import java.io.File

interface ArchiveEngine {
    suspend fun inspect(input: ArchiveInput): ArchiveInspection
    suspend fun readEntry(session: ArchiveEngineSession, path: String, limit: Long, password: String?): ByteArray
    suspend fun extractEntry(session: ArchiveEngineSession, path: String, destination: File, limit: Long, password: String?): Long
    suspend fun createArchive(
        plan: CreationPlan,
        output: ArchiveOutput,
        onProgress: (ProgressSnapshot) -> Unit
    ): CreationResult
}

data class ArchiveInspection(
    val format: String,
    val entryCount: Int,
    val entries: List<EntryDescriptor>,
    val compressedSize: Long?,
    val totalUncompressedSize: Long?,
    val encrypted: Boolean,
    val comment: String?
)

data class EntryDescriptor(
    val index: Int,
    val path: String,
    val kind: String,
    val compressedSize: Long,
    val uncompressedSize: Long,
    val encrypted: Boolean,
    val compressionMethod: String,
    val crc32: UInt? = null,
    val unixMode: UShort? = null
)

data class CreationPlan(
    val entries: List<EntryInput>,
    val compressionProfile: String?,
    val compressionLevel: Int?,
    val storeAlreadyCompressed: Boolean,
    val encryptionMethod: String? = null,
    val encryptionPassword: String? = null
)

data class EntryInput(
    val kind: String,
    val sourcePath: String?,
    val archivePath: String,
    val data: ByteArray?
)

data class CreationResult(
    val entryCount: Int,
    val inputBytes: Long,
    val outputBytes: Long
)
